### PR TITLE
Backend/slide api route implementation

### PR DESCRIPTION
**Pull request by danielsv03**
Added slide API routes and their implementation along with documentation.

### CODE REVIEW

**Review by jonathan-blomlof — APPROVED** [review]

File: Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs [hunk 1]
```diff
@@ -2,7 +2,7 @@ namespace Meta_TV2_api.Controllers;
 
 using Meta_TV2_BusinessLayer;
 using Microsoft.AspNetCore.Mvc;
-
+using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs]: Unnecessary using

File: Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs [hunk 2]
```diff
@@ -43,3 +42,39 @@ public async Task<IActionResult> DeleteGroup(int id)
...
+
+    [HttpGet]
+    public async Task<IActionResult> GetSlides() {
+        var slides = await businessRules.GetSlides();
+        return slides != null ? Ok(slides) : NotFound("No slides found");
+    }
+    [HttpGet("{id}")]
+    public async Task<IActionResult> GetSlidesByGroup(int id) {
+        var slides = await businessRules.GetSlidesByGroup(id);
+        return slides != null ? Ok(slides) : NotFound("No slides found");
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs]: Response when not found could be improved for debugging, consider writing "No slides found for groupId {id}"

File: Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs [hunk 3]
```diff
@@ -43,3 +42,39 @@ public async Task<IActionResult> DeleteGroup(int id)
...
+    }
+
+
+    [HttpGet("{id}/page={page}&size={size}")]
+    public async Task<IActionResult> GetSlidesByGroupPage(int id, int page, int size) {
+        var slides = await businessRules.GetSlidesByGroup(id, page, size);
+        return slides != null ? Ok(slides) : NotFound("No slides found");
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs]: Response when not found could be improved for debugging, consider writing "No slides found for groupId {id}"

File: Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs [hunk 4]
```diff
@@ -43,3 +42,39 @@ public async Task<IActionResult> DeleteGroup(int id)
...
+    }
+
+    [HttpPost]
+    public async Task<IActionResult> CreateSlide(string slideObject) {
+        var created = await businessRules.CreateSlide(slideObject);
+        return created ? Ok() : BadRequest("Failed too add Slide");
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs]: Spelling

File: Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs [hunk 5]
```diff
@@ -43,3 +42,39 @@ public async Task<IActionResult> DeleteGroup(int id)
...
+    }
+
+    [HttpDelete("{id}")]
+    public async Task<IActionResult> ArchiveSlide(int id) {
+        var delete = await businessRules.ArchiveSlide(id);
+        return delete ? Ok() : BadRequest($"Failed to archive slide: {id}");
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs]: 👍

File: Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Program.cs [hunk 6]
```diff
@@ -7,6 +7,8 @@
 builder.Services.AddEndpointsApiExplorer();
 builder.Services.AddSwaggerGen();
 
+
+
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Program.cs:11]: Remove

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs [hunk 7]
```diff
@@ -2,6 +2,7 @@
 
 using System.Text.Json;
 using Meta_TV2_DataLayer;
+using Microsoft.EntityFrameworkCore.Migrations.Operations;
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs]: Unnecessary using

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs [hunk 8]
```diff
@@ -76,4 +77,83 @@ public class BusinessRules : IBusinessRules
...
+        catch (Exception e)
+        {
+            return null;
+        }
+    }
+
+    public async Task<string> GetSlidesByGroup(int groupId) {
+        try {
+            Optional<List<Slides>> result = await DataAccess.GetSlidesByGroup(groupId);
+            if(!result.HasValue) {
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs]: Consider reformatting if-statements like those found in group methods for consistent and clean code. (Applies to all if-statements)

File: Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs [hunk 9]
```diff
@@ -43,3 +42,39 @@ public async Task<IActionResult> DeleteGroup(int id)
...
+    }
+
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs]: Remove

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs [hunk 10]
```diff
@@ -76,4 +77,83 @@ public class BusinessRules : IBusinessRules
             return null;
         }
     }
+
+
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs:80]: Leave only one whitespace

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs [hunk 11]
```diff
@@ -76,4 +77,83 @@ public class BusinessRules : IBusinessRules
...
+            if(!result.HasValue){
+                return null;
+            }
+            return JsonSerializer.Serialize(result.Value);
+        } catch(Exception e) {
+            return null;
+        }
+    }
+
+
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs]: Leave only one whitespace

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs [hunk 12]
```diff
@@ -76,4 +77,83 @@ public class BusinessRules : IBusinessRules
             return null;
         }
     }
+
+
+    public async Task<string> GetSlides() {
+        try
+        {
+            Optional<List<Slides>> result = await DataAccess.GetSlides();
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs]: Consider using "var". (Applies to all slide methods)

File: Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs [hunk 13]
```diff
@@ -43,3 +42,39 @@ public async Task<IActionResult> DeleteGroup(int id)
...
+
+    [HttpGet("{id}/page={page}&size={size}")]
+    public async Task<IActionResult> GetSlidesByGroupPage(int id, int page, int size) {
+        var slides = await businessRules.GetSlidesByGroup(id, page, size);
+        return slides != null ? Ok(slides) : NotFound("No slides found");
+    }
+
+    [HttpPost]
+    public async Task<IActionResult> CreateSlide(string slideObject) {
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-AccessLayer/Controllers/ApiController.cs]: Ambiguity between "CREATEslide" and "ADDgroup". Stick with one. (Remember to change the name in all layers)

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs [hunk 14]
```diff
@@ -76,4 +77,83 @@ public class BusinessRules : IBusinessRules
...
+        }
+    }
+
+    public async Task<bool> CreateSlide(string slideObject){
+        try {
+            var obj = JsonSerializer.Deserialize<Slides>(slideObject);
+            var status = await DataAccess.CreateSlide(obj);
+            return status;
+        } catch(Exception e) {
+            Console.WriteLine(e);
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/BusinessRules.cs]: Logging has not been decided yet, remove.

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/IBusinessRules.cs [hunk 15]
```diff
@@ -43,4 +44,46 @@ public interface IBusinessRules
...
+    /// <param name="id">the slide Id to filter on</param>
+    /// <returns>Slide JSON object if any slide were found, otherwise null</returns>
+    public Task<string> GetSlideById(int id);
+
+    /// <summary>
+    /// Gets slides (non-archived) by pagination by invoking datalayer and serializing result to JSON object
+    /// </summary>
+    /// <param name="groupId"></param>
+    /// <param name="page"></param>
+    /// <param name="size"></param>
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/IBusinessRules.cs]: Missing parameter documentation

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/IBusinessRules.cs [hunk 16]
```diff
@@ -43,4 +44,46 @@ public interface IBusinessRules
...
+    public Task<string> GetSlides();
+
+    /// <summary>
+    /// Gets all slides (non-archived) associated with a specific group Id by invoking datalayer and serializing result to JSON object.
+    /// </summary>
+    /// <param name="groupId">The group Id to filter on</param>
+    /// <returns>Slide JSON object if any slides were found, otherwise null</returns>
+    public Task<string> GetSlidesByGroup(int groupId);
+    /// <summary>
+    /// Gets a specific slide (can be archived) associated with a specific slide Id by invoking datalayer and serializing result to JSON object.
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/IBusinessRules.cs]: Unclear, consider changing to something like "(including archived)"

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/IBusinessRules.cs [hunk 17]
```diff
@@ -43,4 +44,46 @@ public interface IBusinessRules
...
+    /// Gets slides (non-archived) by pagination by invoking datalayer and serializing result to JSON object
+    /// </summary>
+    /// <param name="groupId"></param>
+    /// <param name="page"></param>
+    /// <param name="size"></param>
+    /// <returns>Slide JSON object if any slide were found, otherwise null<</returns>
+    public Task<string> GetSlidesByGroup(int groupId, int page, int size);
+
+    /// <summary>
+    /// Deserializes a Slide JSON object and adds the desralized object to the database by invoking datalayer.
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/IBusinessRules.cs]: Spelling

File: Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/IBusinessRules.cs [hunk 18]
```diff
@@ -43,4 +44,46 @@ public interface IBusinessRules
...
+    /// </summary>
+    /// <param name="slideObject">Slide JSON object</param>
+    /// <returns>True if operation successfull otherwise false</returns>
+    public Task<bool> CreateSlide(string slideObject);
+
+    /// <summary>
+    /// Archives a specific slide in the database by updating slide archived field and invoking datalayer.
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-BusinessLayer/IBusinessRules.cs]: "fields"?

File: Backend/Meta-TV2-api/Meta-TV2-DataLayer/DataAccess.cs [hunk 19]
```diff
@@ -1,4 +1,5 @@
-﻿using Microsoft.EntityFrameworkCore;
+﻿using System.Data.Common;
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-DataLayer/DataAccess.cs]: Unnecessary using

File: Backend/Meta-TV2-api/Meta-TV2-DataLayer/DataAccess.cs [hunk 20]
```diff
@@ -62,4 +63,58 @@ public class DataAccess : IDataAccess
...
+        var query = from x in db.Slides where x.archive == false & x.groupId == groupId select x;
+        List<Slides> groups = await query.Skip((page-1) * size).Take(size).ToListAsync();
+        if (groups.Count == 0){
+            return Optional<List<Slides>>.Empty();
+        }
+        return Optional<List<Slides>>.Result(groups);
+    }
+
+    public async Task<bool> CreateSlide(Slides obj) {
+        try{
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-DataLayer/DataAccess.cs]: What could cause an exception? Entity classes are fixed to force all necessary fields to be filled in (in another branch).

File: Backend/Meta-TV2-api/Meta-TV2-DataLayer/DataAccess.cs [hunk 21]
```diff
@@ -62,4 +63,58 @@ public class DataAccess : IDataAccess
...
+            db.Add(obj);
+            await db.SaveChangesAsync();
+            db.Dispose();
+            return true;
+        } catch (Exception e) {
+            Console.WriteLine(e);
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-DataLayer/DataAccess.cs]: Logging has not been decided yet, remove.

File: Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs [hunk 22]
```diff
@@ -46,5 +46,50 @@ public interface IDataAccess
...
+    /// <summary>
+    /// Gets a specific slide associated with a specific slide Id from the databse (ignores if the slide is archived or not)
+    /// </summary>
+    /// <param name="id">the slide Id to filter against</param>
+    /// <returns>Optional object with a slide object, otherwise empty Optional object if no slide was found</returns>
+    public Task<Optional<Slides>> GetSlideById(int id);
+
+    /// <summary>
+    /// Gets all non-archived slides associated with a specific group Id, page and size from the database.
+    /// The size determens the how many slides to return, and page determens which set of slides to return.
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs]: Spelling "determines" (appears twice). Remove "the" in front of the first "determens"

File: Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs [hunk 23]
```diff
@@ -46,5 +46,50 @@ public interface IDataAccess
...
+    /// Page is 1 indexed.
+    /// </summary>
+    /// <param name="groupId">the group Id to filter slides on</param>
+    /// <param name="page">which set of result to return</param>
+    /// <param name="size">the size of the result</param>
+    /// <returns>Optional object witha list of slide objects as value, otherwise empty Optional object if no slides found</returns>
+    public Task<Optional<List<Slides>>> GetSlidesByGroup(int groupId, int page, int size);
+
+    /// <summary>
+    /// Creates a new Slide object and stores the Slide object inside the database.
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs]: Does the method really "create a new Slide object?"

File: Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs [hunk 24]
```diff
@@ -46,5 +46,50 @@ public interface IDataAccess
...
+    /// </summary>
+    /// <param name="obj">The slide object to store</param>
+    /// <returns>True if operation was sucessfull, False if not</returns>
+    public Task<bool> CreateSlide(Slides obj);
+
+    /// <summary>
+    /// Updates a already existing slide in the databse and stores the update
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs]: "an"

File: Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs [hunk 25]
```diff
@@ -46,5 +46,50 @@ public interface IDataAccess
...
+    /// Will update the slide with the given attributes based on the slideId given inside of (slide)
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs]: The newline is not represented when previewing documentation from the code implementation. Consider using a period. Remove "... given inside of slide"

File: Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs [hunk 26]
```diff
@@ -46,5 +46,50 @@ public interface IDataAccess
...
+    /// <param name="size">the size of the result</param>
+    /// <returns>Optional object witha list of slide objects as value, otherwise empty Optional object if no slides found</returns>
+    public Task<Optional<List<Slides>>> GetSlidesByGroup(int groupId, int page, int size);
+
+    /// <summary>
+    /// Creates a new Slide object and stores the Slide object inside the database.
+    /// </summary>
+    /// <param name="obj">The slide object to store</param>
+    /// <returns>True if operation was sucessfull, False if not</returns>
+    public Task<bool> CreateSlide(Slides obj);
```
Comment (DanielNikolic03) [inline Backend/Meta-TV2-api/Meta-TV2-DataLayer/IDataAccess.cs]: Consider calling it "slide" instead of "obj"

**Review by DanielNikolic03 — CHANGES_REQUESTED** [review]

**Review by DanielNikolic03 — APPROVED** [review]

**Pull request merged**